base/src/main/java/com/thoughtworks/go/util/ZipUtil.java

```diff
@@ -104,8 +104,8 @@ void addToZip(ZipPath path, File srcFile, ZipOutputStream zip, boolean excludeRo
         }
     }
 
-    private void bombIfZipEntryPathContainsDirectoryTraversalCharacters(String filepath) {
-        if (filepath.contains("..")) {
+    private void failIfZipEntryPathContainsDirectoryTraversalCharacters(String filepath) {
+        if (filepath.contains("/../") || filepath.startsWith("../") || filepath.endsWith("/..")) {
```

Comment:
Didn't want to actually get the canonical path and compare, etc. Thought it might just add a bit of overhead.

Reply:
What about a windows equivalent path `\..\` or `\\..\\` depending on whether it's escaped, or not?

`FileNameUtils#separatorsToUnix(filepath).split('/').contains("..")` is what I'd think we'd want?

Reply:
@arvindsv - why not make use of something like `FilenameUtil.isNormalizedPathOutsideWorkingDir` ? Its what we use to validate that the task working directory is not outside the agent sandbox.

Reply:
@varshavaradarajan I did consider using something like:

```
filepath.contains("..") && FilenameUtil.isNormalizedDirectoryPathInsideNormalizedParentDirectory(dir, filepath)
```

but it seems unnecessary. A zip entry (an entry *inside* a zip file) with `../` in the name seems unnecessary to support. Also, `isNormalizedDirectoryPathInsideNormalizedParentDirectory` expects the second argument to be a directory - and changing/rewriting that seemed unnecessary. What do you think?

Reply:
@ketan I think you're right. I'll think about it. In this case though, this is a zip entry. The backslash separator doesn't really apply. However, I'll need to see how such a zip entry shows up in Windows. Meaning, whether a zip file created in Windows uses backslash or it is standardized to a forward slash.

Reply:
@ketan FYI:

- The [zip format specification](https://pkware.cachefly.net/webdocs/casestudies/APPNOTE.TXT) says that the slashes must be forward slashes only.
- .Net framework was also finally changed in May this year to reflect this.

I guess it is possible for zip files generated on Windows (and maybe by our Windows agents) to have backslash as the name separator. I'll verify and convert them to a forward slash before comparison, if that's the case.

Reply:
@adityasood @ketan I tried this with a Windows agent too and it works. The zip file is generated and uploaded (and downloaded on the server successfully).

Also, tried generating [a zip file](https://gist.github.com/arvindsv/89b334567e9ddebd43aaf70cca88e704) with Windows separators and a `\..\` and it uploaded to a server running on Mac OS X and it created a file with backslashes in it, which is valid on a Mac.

```
$ unzip -l windows_dots.zip
Archive:  windows_dots.zip
  Length      Date    Time    Name
---------  ---------- -----   ----
        9  11-19-2018 18:43   1.txt
        9  11-19-2018 18:43   \..\2.txt
---------                     -------
       18                     2 files

$ curl http://localhost:8153/go/files/Test-Zipping/2/defaultStage/2/defaultJob/test2 -H 'Confirm: true' -X POST -F 'zipfile=@windows_dots.zip'
File test2 was created successfully

$ ls -l
total 8,192
-rw-r--r-- 1 user staff 9 Nov 19 18:53  1.txt
-rw-r--r-- 1 user staff 9 Nov 19 18:53 '\..\2.txt'
```

Just to be on the safe side, I'll look at disallowing forward and backslashes, disregarding which OS this code is running on (since it could be a Windows agent + Linux server or Linux agent + Windows server ... etc.)